cmd/occollector/app/builder/exporters_builder.go

```diff
@@ -0,0 +1,252 @@
+	case configmodels.MetricsDataType:
+		return configv2.MetricsDataTypeStr
+	default:
+		panic("unknown data type")
+	}
+}
+
+// combineStopFunc combines 2 functions and returns one function
+// that can be called and which in turn will call both functions
+// and then combine any errors that the 2 functions return.
+// Safe to use if any of the 2 functions are nil. If both functions
+// are nil then returns nil.
+func combineStopFunc(f1, f2 factories.StopFunc) factories.StopFunc {
+	if f1 == nil {
+		return f2
+	}
+	if f2 == nil {
+		return f1
+	}
+
+	return func() error {
+		var errs []error
+		if err := f1(); err != nil {
+			errs = append(errs, err)
+		}
+		if err := f2(); err != nil {
+			errs = append(errs, err)
+		}
+		return internal.CombineErrors(errs)
+	}
+}
+
+func (eb *ExportersBuilder) buildExporter(
+	config configmodels.Exporter,
+	exportersInputDataTypes exportersRequiredDataTypes,
+) (*exporterImpl, error) {
+
+	factory := factories.GetExporterFactory(config.Type())
+
+	exporter := &exporterImpl{}
+
+	inputDataTypes := exportersInputDataTypes[config]
+	if inputDataTypes == nil {
+		// No data types where requested for this exporter. This can only happen
+		// if there are no pipelines associated with the exporter.
+		eb.logger.Info("Exporter " + config.Name() +
+			" is not associated with any pipeline and will not export data.")
+		return exporter, nil
+	}
+
+	if requirement, ok := inputDataTypes[configmodels.TracesDataType]; ok {
+		// Traces data type is required. Create a trace exporter based on config.
+		tc, stopFunc, err := factory.CreateTraceExporter(config)
+		if err != nil {
+			if err == factories.ErrDataTypeIsNotSupported {
+				// Could not create because this exporter does not support this data type.
+				return nil, typeMismatchErr(config, requirement.requiredBy, configmodels.TracesDataType)
+			}
+			return nil, fmt.Errorf("error creating %q exporter: %v", config.Name(), err)
+		}
+
+		exporter.tc = tc
+		exporter.stop = stopFunc
+	}
+
+	if requirement, ok := inputDataTypes[configmodels.MetricsDataType]; ok {
+		// Metrics data type is required. Create a trace exporter based on config.
+		mc, stopFunc, err := factory.CreateMetricsExporter(config)
+		if err != nil {
+			if err == factories.ErrDataTypeIsNotSupported {
+				// Could not create because this exporter does not support this data type.
+				return nil, typeMismatchErr(config, requirement.requiredBy, configmodels.MetricsDataType)
+			}
+			return nil, fmt.Errorf("error creating %q exporter: %v", config.Name(), err)
+		}
+
+		exporter.mc = mc
+		exporter.stop = combineStopFunc(exporter.stop, stopFunc)
+	}
+
+	return exporter, nil
+}
+
+func typeMismatchErr(
+	config configmodels.Exporter,
+	requiredByPipeline *configmodels.Pipeline,
+	dataType configmodels.DataType,
+) error {
+	return fmt.Errorf("pipeline %q produces %q to exporter "+
+		"%s which does not support %q telemetry data. "+
+		"exporter will be detached from pipeline", requiredByPipeline.Name,
```

Comment:
nit: `requiredByPipeline.Name` should be on the next line together with the other parameters 
(or each parameter on its own line, except format string)